cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java

```diff
@@ -241,20 +252,42 @@ public void deleteNamespace(Id.Namespace namespaceId) throws DatasetManagementEx
     clientCache.getUnchecked(namespaceId).deleteNamespace();
   }
 
-  private void addModule(Id.DatasetModule moduleId, Class<?> typeClass) throws DatasetManagementException {
-    try {
-      File tempFile = File.createTempFile(typeClass.getName(), ".jar");
+  private void addModule(Id.DatasetModule moduleId, Class<?> typeClass,
+                         @Nullable Location jarLocation) throws DatasetManagementException {
+
+    Location deploymentJar = jarLocation;
+    // Only trace dependency and create new jar if jar location is not provided.
+    if (deploymentJar == null) {
```

Comment:
may be a little cleaner to have new jar creation and cleanup logic in the other method to avoid null checks.